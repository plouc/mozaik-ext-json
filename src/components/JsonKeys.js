import React, { Fragment, Component } from 'react'
import PropTypes from 'prop-types'
import _ from 'lodash'
import {
    TrapApiError,
    Widget,
    WidgetHeader,
    WidgetBody,
    WidgetLoader,
    WidgetListItem,
} from '@mozaik/ui'
import computeRequestId from '../lib/computeRequestId'

export default class JsonKeys extends Component {
    static propTypes = {
        title: PropTypes.string,
        url: PropTypes.string.isRequired,
        headers: PropTypes.object,
        keys: PropTypes.arrayOf(PropTypes.string).isRequired,
        apiData: PropTypes.object,
        apiError: PropTypes.object,
    }

    static getApiRequest({ url, headers = {} }) {
        return {
            id: computeRequestId(url, headers),
            params: { url, headers },
        }
    }

    render() {
        const { title, keys, apiData, apiError } = this.props

        let body = <WidgetLoader />
        if (apiData && !apiError) {
            body = (
                <Fragment>
                    {keys.map(key => {
                        let value = _.get(apiData, key)
                        if (_.isBoolean(value)) {
                            value = value ? 'true' : 'false'
                        }

                        return (
                            <WidgetListItem key={key} title={key} post={<strong>{value}</strong>} />
                        )
                    })}
                </Fragment>
            )
        }

        const hasHeader = title && title.length > 0

        return (
            <Widget>
                {hasHeader && <WidgetHeader title={title} />}
                <WidgetBody disablePadding={true} isHeaderless={hasHeader}>
                    <TrapApiError error={apiError}>{body}</TrapApiError>
                </WidgetBody>
            </Widget>
        )
    }
}
