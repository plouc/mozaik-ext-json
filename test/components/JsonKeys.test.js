import React from 'react'
import Enzyme from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import renderer from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'
import { WidgetHeader, WidgetLoader, WidgetListItem, defaultTheme } from '@mozaik/ui'
import JsonKeys from './../../src/components/JsonKeys'

Enzyme.configure({ adapter: new Adapter() })

describe('getApiRequest', () => {
    it('should return correct api request', () => {
        expect(
            JsonKeys.getApiRequest({
                url: 'http://fake.io',
            })
        ).toEqual({
            id: `json.request.http://fake.io.`,
            params: {
                url: 'http://fake.io',
                headers: {},
            },
        })
    })

    it('should support additional headers', () => {
        expect(
            JsonKeys.getApiRequest({
                url: 'http://fake.io',
                headers: {
                    'X-Whatever': 'yay',
                },
            })
        ).toEqual({
            id: `json.request.http://fake.io.X-Whatever:yay`,
            params: {
                url: 'http://fake.io',
                headers: {
                    'X-Whatever': 'yay',
                },
            },
        })
    })
})

it('should display loader if no apiData available', () => {
    const wrapper = Enzyme.shallow(<JsonKeys url="http://fake.io" keys={['A']} />)

    expect(wrapper.find(WidgetLoader).exists()).toBeTruthy()
})

describe('title', () => {
    it('should display no header by default', () => {
        const wrapper = Enzyme.shallow(<JsonKeys url="http://fake.io" keys={['A']} />)

        expect(wrapper.find(WidgetHeader).exists()).toBeFalsy()
    })

    it('should support override', () => {
        const wrapper = Enzyme.shallow(
            <JsonKeys url="http://fake.io" keys={['A']} title="override" />
        )

        const header = wrapper.find(WidgetHeader)
        expect(header.prop('title')).toBe('override')
    })
})

it('should display configured key values', () => {
    const wrapper = Enzyme.shallow(
        <JsonKeys
            url="http://fake.io"
            keys={['A', 'B', 'C']}
            apiData={{
                A: 'A value',
                B: 'B value',
                C: 'C value',
            }}
        />
    )

    const items = wrapper.find(WidgetListItem)
    expect(items).toHaveLength(3)
    expect(items.at(0).prop('title')).toBe('A')
    expect(items.at(0).prop('post').props.children).toBe('A value')
    expect(items.at(1).prop('title')).toBe('B')
    expect(items.at(1).prop('post').props.children).toBe('B value')
    expect(items.at(2).prop('title')).toBe('C')
    expect(items.at(2).prop('post').props.children).toBe('C value')
})

it('should support boolean values', () => {
    const wrapper = Enzyme.shallow(
        <JsonKeys
            url="http://fake.io"
            keys={['A']}
            apiData={{
                A: true,
            }}
        />
    )

    const items = wrapper.find(WidgetListItem)
    expect(items).toHaveLength(1)
    expect(items.at(0).prop('title')).toBe('A')
    expect(items.at(0).prop('post').props.children).toBe('true')
})

it('should render as expected', () => {
    const tree = renderer.create(
        <ThemeProvider theme={defaultTheme}>
            <JsonKeys
                url="http://fake.io"
                keys={['A', 'B', 'C']}
                apiData={{
                    A: 'A value',
                    B: 'B value',
                    C: 'C value',
                }}
            />
        </ThemeProvider>
    )

    expect(tree).toMatchSnapshot()
})
