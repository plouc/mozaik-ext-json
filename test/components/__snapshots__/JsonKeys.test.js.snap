// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render as expected 1`] = `
<div
  className="sc-bdVaJa eydJRb"
  style={Object {}}
>
  <div
    className="sc-bwzfXH bJUZVN"
  >
    <div
      className="sc-bZQynM FCrnA"
      style={Object {}}
    >
      <div
        className="sc-dnqmqq Bpvft"
      >
        <div
          style={
            Object {
              "flexGrow": 1,
            }
          }
        >
          <div
            className="sc-iwsKbI gExwvp"
          >
            A
          </div>
        </div>
        <div
          className="sc-gqjmRU cABAvw"
        >
          <strong>
            A value
          </strong>
        </div>
      </div>
      <div
        className="sc-dnqmqq Bpvft"
      >
        <div
          style={
            Object {
              "flexGrow": 1,
            }
          }
        >
          <div
            className="sc-iwsKbI gExwvp"
          >
            B
          </div>
        </div>
        <div
          className="sc-gqjmRU cABAvw"
        >
          <strong>
            B value
          </strong>
        </div>
      </div>
      <div
        className="sc-dnqmqq Bpvft"
      >
        <div
          style={
            Object {
              "flexGrow": 1,
            }
          }
        >
          <div
            className="sc-iwsKbI gExwvp"
          >
            C
          </div>
        </div>
        <div
          className="sc-gqjmRU cABAvw"
        >
          <strong>
            C value
          </strong>
        </div>
      </div>
    </div>
  </div>
</div>
`;
